L1Trigger/L1TTrackMatch/plugins/L1TkMuonProducer.cc

```diff
@@ -251,105 +316,44 @@ L1TkMuonProducer::runOnMTFCollection_v1(const edm::Handle<RegionalMuonCandBxColl
-
-      float l1tk_eta = l1tk.getMomentum(nPars).eta();
-      float l1tk_phi = l1tk.getMomentum(nPars).phi();
-
-      float dr2 = deltaR2(l1mu_eta, l1mu_phi, l1tk_eta, l1tk_phi);
-      if (dr2 > 0.3) continue;
-
-      const PropState& pstate = propagateToGMT(l1tk);
-      if (!pstate.valid) continue;
-
-      float dr2prop = deltaR2(l1mu_eta, l1mu_phi, pstate.eta, pstate.phi);
-      // FIXME: check if this matching procedure can be improved with
-      // a pT dependent dR window
-      if (dr2prop < drmin){
-        drmin = dr2prop;
-        match_idx = il1tk;
-        matchProp = pstate;
-      }
-    }// over l1tks
-
-    LogDebug("MYDEBUG")<<"matching index is "<<match_idx;
-    if (match_idx >= 0){
-      const L1TTTrackType& matchTk = l1tks[match_idx];
-
-      // float etaCut = 3.*sqrt(l1mu->hwDEtaExtra()*l1mu->hwDEtaExtra() + matchProp.sigmaEta*matchProp.sigmaEta);
-      // float phiCut = 4.*sqrt(l1mu->hwDPhiExtra()*l1mu->hwDPhiExtra() + matchProp.sigmaPhi*matchProp.sigmaPhi);
-
-      // float dEta = std::abs(matchProp.eta - l1mu->eta());
-      // float dPhi = std::abs(deltaPhi(matchProp.phi, l1mu->phi()));
-
-      // LogDebug("MYDEBUG")<<"match details: prop "<<matchProp.pt<<" "<<matchProp.eta<<" "<<matchProp.phi
-			//  <<" mutk "<<l1mu->pt()<<" "<<l1mu->eta()<<" "<<l1mu->phi()<<" delta "<<dEta<<" "<<dPhi<<" cut "<<etaCut<<" "<<phiCut;
-      if (drmin < DRmax_){
-        edm::Ptr< L1TTTrackType > l1tkPtr(l1tksH, match_idx);
-
-        unsigned int nPars = 4;
-        if (use5ParameterFit_) nPars = 5;
-        const auto& p3 = matchTk.getMomentum(nPars);
-        float p4e = sqrt(0.105658369*0.105658369 + p3.mag2() );
-
-        math::XYZTLorentzVector l1tkp4(p3.x(), p3.y(), p3.z(), p4e);
-
-        const auto& tkv3=matchTk.getPOCA(nPars);
-        math::XYZPoint v3(tkv3.x(), tkv3.y(), tkv3.z());
-        float trkisol = -999;
+  const EMTFTrackCollection& l1mus = (*muonH.product());
+  const L1TTTrackCollectionType& l1trks = (*l1tksH.product());
+  auto corr_mu_idxs = dwcorr_->find_match(l1mus, l1trks);
+  // it's a vector with as many entries as the L1TT vector.
+  // >= 0 : the idx in the EMTF vector of matched mu
+  // < 0: no match
+
+  // sanity check
+  if (corr_mu_idxs.size() != l1trks.size())
+    throw cms::Exception("TkMuAlgoOutput") << "the size of tkmu indices does not match the size of input trk collection\n";
+  
+  for (uint il1ttrack = 0; il1ttrack < corr_mu_idxs.size(); ++il1ttrack)
+  {
+    int emtf_idx = corr_mu_idxs.at(il1ttrack);
+    if (emtf_idx < 0)
+      continue;
+
+    const L1TTTrackType& matchTk = l1trks[il1ttrack];
+    const auto& p3 = matchTk.getMomentum(dwcorr_->get_n_trk_par());
+    const auto& tkv3 = matchTk.getPOCA(dwcorr_->get_n_trk_par());
+    float p4e = sqrt(0.105658369*0.105658369 + p3.mag2() );
+    math::XYZTLorentzVector l1tkp4(p3.x(), p3.y(), p3.z(), p4e);
+
+    edm::Ref< RegionalMuonCandBxCollection > l1muRef(nullptr, 0); // FIXME! The reference to the muon is null
```

Comment:
This is bad, since you're creating a null **transient** reference, and so CMSSW will throw an exception if one tries to save the object to an EDM file (error message: `Inconsistency RefCoreStreamer: transient Ref or Ptr cannot be made persistent.`)

The proper way to create a null reference is to just use the default constructor, i.e. just do instead
````edm::Ref< RegionalMuonCandBxCollection > l1muRef;````
and things should work fine (in particular, with that change, I see that I can now save the TkMuons in an EDM file without errors)

it seems to me #715 doesn't change this code, so I'll make a bugfix PR for this.